spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/RetryGatewayFilterFactory.java

```diff
@@ -153,13 +161,57 @@ public void reset(ServerWebExchange exchange) {
 		exchange.getAttributes().remove(GATEWAY_ALREADY_ROUTED_ATTR);
 	}
 
+	public Mono<Void> preparedRequest(ServerWebExchange exchange,
+			GatewayFilterChain chain) {
+		return Mono.fromCallable(() -> {
+			Optional<DataBuffer> cachedRequestBuffer = exchange
```

Comment:
Why use optional here? I think it's fairly minor (ie, doesn't affect actual functionality, but looks strange). We avoid it here https://github.com/spring-cloud/spring-cloud-gateway/blob/master/spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/handler/predicate/ReadBodyPredicateFactory.java

Reply:
If I do not use optional - it would have to be refactored to handle the empty case e.g. `switchIfEmpty` instead of `defaultIfEmpty`.

Using the optional allows to do this in a single observable.  I can go the other way if you prefer that.

In what you linked, there are specific comments.

```
// Note: if we end up reading the body twice we will run into
// a problem, but as of right
// now there is no good use case for doing this
```

There might be some issues in that code specifically around retry as well.  I believe it would leak memory.

Reply:
The above code as well, I don't believe will do anything if the request body is empty i.e. your standard GET.

Reply:
:+1: thanks for the good explanation.